Enable TRY_CONVERT Transact-SQL Function

Jakeowen1 opened the pull request:
Previously the TRY_CONVERT function was unsupported by Babelfish. This commit enables the TRY_CONVERT Transact-SQL function. To enable TRY_CONVERT this commit adds the babelfish_conv_string_to_time function
to the upgrade scripts along with associated validation tests.

Signed-off-by: Jake Owen <owjco@amazon.com>
 
### Issues Resolved

Babel-3486

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Enable TRY_CONVERT Transact-SQL Function

Previously the TRY_CONVERT function was unsupported by Babelfish. This
commit enables the TRY_CONVERT Transact-SQL function. To enable
TRY_CONVERT this commit adds the babelfish_conv_string_to_time function
to the upgrade scripts along with associated validation tests.

Task: BABEL-3486
Signed-off-by: Jake Owen <owjco@amazon.com>

shah-nirmit committed:
Change SUM aggregates for integer datatypes to return correct data types (#563)

Problem: The SUM aggregates were having wrong return datatypes

Fix Approach: New final functions were introduced to return correct datatype and  sum(bigint),sum(int),sum(smallint),sum(int) aggregates were recreated from babelfishpg_common extension  so that these aggregates are given more precedance than pg_catalog.sum

Issues Resolved: BABEL-3144
Engine PR: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/47
Authored-by: Nirmit Shah <nirmisha@amazon.com>
Signed-off-by: Dipesh Dhameliya <dddhamel@amazon.com>

sumitj824 committed:
Add support for all multi-byte currency symbols

Earlier Babel only supported dollar currency symbol because other currency symbols are multi-byte characters because in the parser we had support only for single byte currency symbols. With this commit we have modified the parser in order to support other currency symbols.
 
Task: BABEL-2636
authored-by: Sumit Jaiswal <sumiji@amazon.com>
Signed-off-by: Kushaal Shroff <kushaal@amazon.com>

ahmed-shameem committed:
Adding test cases for invalid CREATE TABLE and column name (#582)

Adding test cases for invalid CREATE TABLE and column name
 
Issues Resolved: BABEL-1953

Tests scenarios covered:

- Negative test scenarios

    o	Tests added for table creation without a table name
    o	Tests added for table creation without datatype of a column
    o	Tests added for table creation without column name but with datatypes
    o	Tests added for invalid table name
    o	Tests added for invalid column name
    o	Tests added for table without column
    o	Tests added for views which include invalid view name




Signed-off-by: Shameem Ahmed <shameemahmed20apr2000@gmail.com>

basasairohan committed:
Added upgrade tests for system views in 14.5 and latest schedule files (#622)

 Added upgrade tests for system views in 14.5 and latest schedule files

raydhkim committed:
Fix IS_ROLEMEMBER(), IS_MEMBER() ignoring Trailing Spaces in arguments

Previously, Babelfish did not ignore trailing spaces in the input for functions: is_rolemember() and is_member(). Now, it trims the unnecessary trailing spaces of the inputs, so that is_rolemember() and is_member() work as the inputs do not have the spaces.

Task: BABEL-3456
Signed-off-by: Ray Kim <raydhkim@amazon.com>

wboettge committed:
Add sp_helpdbfixedrole

Adds the sp_helpdbfixedrole stored procedure. The procedure takes a name
of a fixed database role, and returns information about that role. If no
role is specified return a list and description of all fixed database
roles.

Currently, db_owner is the only fixed database-level role from
SQL-Server in Babelfishdb_owner. Like sp_helpsrvrolemember in
BABEL-3462, an empty result set will be returned if the procedure is
called with any of the other 8 unsupported fixed db roles.

If called with any input besides one of SQL-Server's fixed db roles, an
error is raised.

Task: BABEL-3086

Signed-off-by: Walt Boettge <wboettge@amazon.com>

ZhenyeLee committed:
Fix error in CONVERT using character string into DATE or DATETIME (#621)

* Fix error in CONVERT using character string into DATE or DATETIME

Previously, CONVERT function fails when character string of format mm/dd/yy convert into DATE or DATETIME. The error is :operator is not unique: smallint > numeric. The error raised in ``babelfish_get_full_year`` function when trying to compare SMALLINT with NUMERIC. In this commit,  SMALLINT is converted into NUMERIC before doing comparison.

Task:BABEL-3314
Signed-off-by: Zhenye Li <zhenye@amazon.com>

AvantikaDG commented on test/JDBC/expected/BABEL-3486-vu-prepare.out:
```diff
@@ -0,0 +1,161 @@
+CREATE PROCEDURE BABEL_3486_vu_prepare_p5 as (SELECT TRY_CONVERT(float, 11234561231231.234));
+GO
+
+CREATE PROCEDURE BABEL_3486_vu_prepare_p6 as (SELECT TRY_CONVERT(MONEY, 55.35));
```
What happens if we have >4 digits after the decimal point? Something like 
`SELECT TRY_CONVERT(MONEY, 55.35791)`.

Let's add a test for that.

AvantikaDG commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -114,6 +114,7 @@ BABEL-2944
 BABEL-3116
 BABEL-3118
 BABEL-3249
+BABEL-3486
```
Can you specify in the commit message that TRY_CONVERT was incorrectly marked as an unsupported feature? Also please update your PR message with the new commit message.

AvantikaDG commented on test/JDBC/input/BABEL-3486-vu-prepare.sql:
```diff
@@ -0,0 +1,161 @@
+CREATE PROCEDURE BABEL_3486_vu_prepare_p3 as (SELECT TRY_CONVERT(time, '13:01:59'));
+GO
+
+CREATE PROCEDURE BABEL_3486_vu_prepare_p4 as (SELECT TRY_CONVERT(datetime, '2017-08-25 13:01:59'));
```
Let's add a test for datetime conversion where the date is not specified
` SELECT TRY_CONVERT(datetime, '13:01:59')`

AvantikaDG commented on test/JDBC/input/BABEL-3486-vu-prepare.sql:
```diff
@@ -0,0 +1,161 @@
+CREATE VIEW BABEL_3486_vu_prepare_v6 as (SELECT TRY_CONVERT(MONEY, 55.35));
+GO
+
+CREATE VIEW BABEL_3486_vu_prepare_v7 as (SELECT TRY_CONVERT(float, 1));
```
Have we covered all the datatypes? Can think of a few more off the top of my head - numeric, tinyint, sql_variant. I think we should look up supported datatypes and cover the following cases on each of them:
1. Valid operation
2. Invalid operation
3. Edge case

rishabhtanwar29 committed:
Fix precision error when sending numeric zero (#608)

* Fix precision error when sending numeric zero

When TDS gets a numeric zero with typmod = -1 to sent to client
then it calls *numeric_get_typmod* function to calculate typmod,
precision and scale using actual expression but numeric zero
has both precision and scale equal to zero so we send 0 to
client directly which causes error on some clients as precision
0 is invalid.

Fix this by sending tds_default_numeric_precision/scale instead
for numeric zero.

Task: BABEL-3344
Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>

* Fix default precision/scale

Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>

* Fix test

Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>

* Add test for non numeric case

Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>

* Trigger tests

Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>

Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>
Co-authored-by: Rishabh Tanwar <ritanwar@amazon.com>

shah-nirmit committed:
Fix github workflows to point to correct branches (#641)

Github actions should point to the stable branches and the latest dev branches for running the upgrade tests.

ZhenyeLee committed:
Add 'w' unit to four date functions (#501)

* Add 'w' unit to four date funtions

Add 'w' unit to DAREPART(), DATENAME(), DATEDIFF(), DATEADD() functions
by add 'w' as a keywork.

Task: BABEL_3358
Signed-off-by: Zhenye Li <zhenye@amazon.com>

* Correct dateiff_arg

Signed-off-by: Zhenye Li <zhenye@amazon.com>

* Correct datepart argments

Signed-off-by: Zhenye Li <zhenye@amazon.com>

* retrigger checks

* Update gram-tsql-rule.y

* Delete tests from upgrade schedule files

Delete tests from upgrade schedule files and update
expected_dependency.out

Signed-off-by: Zhenye Li <zhenye@amazon.com>

Signed-off-by: Zhenye Li <zhenye@amazon.com>
Co-authored-by: Zhenye Li <zhenye@amazon.com>

jusjosj committed:
Expand explain for PRINT, THROW, USE and RAISERROR (#561)

This commit extends the functionality of babelfish_showplan_all to include print, throw, raiserror, and use
This is largely done in a manner consistent with the other explain functions existing throughout the codebase.

Task: BABEL-2902
Signed-off-by: Justin Jossick <jusjosj@amazon.com>

shah-nirmit committed:
Change github actions to point to correct engine dev branch (#649)

Authored-by: Nirmit Shah <nirmisha@amazon.com>
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>

Sairakan committed:
Fix table hint parsing (#637)

Previously, Babelfish would sometimes incorrectly parse a query that used a table alias with column definitions as a table hint and raise an error. This commit fixes the issue by having the parser only accept a specific defined list of keywords as possible table hints.

N.B. Babelfish will report incorrect table hint syntax as generic syntax errors instead of specifically table hint errors as SQL Server does. This commit does not attempt to mimic this behavior, as that would require a more extensive modification to the antlr parser (most likely requiring a new grammar rule to be added).

Task: BABEL-2805
Signed-off-by: Jason Teng jasonten@amazon.com

AmithKamath committed:
Add JDBC tests for CONVERT()

An issue BABEL-2054 was reported where CONVERT(datetime, string) failed
intermittently. However, this issue is not reproducible and this commit
adds JDBC tests to validate the scenario mentioned in the JIRA

Task: BABEL-2054

Signed-off-by: Amith Kamath <amithkam@amazon.com>

wboettge committed:
Login who owns database should always be dbo also without sysadmin role

In SQL Server, the login that creates a database is automatically the
database owner. This means that the login can always access the
databases they create.

However, previously in babelfish, if a login is stripped of the sysadmin
role after creating a database, and the login does not have a
user associated with the database, then the login can no longer access
the database. Now, when a login attempts to access and there
is not an associated user, the login will be considered the dbo and
allowed access.

Task: BABEL-3548

Signed-off-by: Walt Boettge <wboettge@amazon.com>

shah-nirmit committed:
Refactor Encoding Conversion Related Functions to improve Performance (#632)

Previously, we were using pg_any_to_server() API on receiving side for varchar, char and text datatypes.
pg_any_to_server() internally do the sys cache lookup which was expensive in some cases(varchar,char and text datatypes) and we really do not need to do it as TDS only supports collations which requires only few set of encodings. This commit adds new function encoding_conv_util() which would take src_encoding and dest_encoding as argument and will call appropriate encoding conversion functionality based on those arguments. This commit also copies few encoding conversion related functionality (TsqlLocalToUtf)from engine to common extension and extends them to return byte length of encoded string. 
 
Changes: 
1.  Remove duplicate encoding logic from babelfishpg_tsql extension
2. Introduce conditional handling for encoding conversion for Varchar, Char and Text datatypes on receiver side
3. Copy and enhance TsqlLocalToUtf from engine side to be used by various encoding functions


### Testing
- Collation TestFramework
  - Tested Collations
    - sql_latin1_general_cp1_cs_as
    - arabic_cs_as
    - greek_cs_as
    - turkish_cs_as
    - polish_cs_as
    - vietnamese_cs_as
    - ukrainian_cs_as
    - japanese_cs_as
    - korean_wansung_cs_as
- Manual testing with custom C# program

Task: BABEL-3542
Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

ZhenyeLee committed:
Fix syscoumns test failure (#660)

The helper function in prepare file searching object_name by oid in both pg_class and pg_proc. However, pg_class and pg_proc may have the same oid and it cause test failure. In this PR, we break down the search into to functions, one is in pg_class, and another one is in pg_proc. And also removed order by id from verification file, because oid can be changed after MVU.

Signed-off-by: Zhenye Li zhenye@amazon.com

raydhkim committed:
Fix: Unable to connect through SSMS - When Create Login with sysadmin role

Previously, when user created a login with sysadmin role and connected through SSMS, permission error was thrown. Now, user can successfully connect using SSMS with a login with sysadmin role.

Task: BABEL-3550
Signed-off-by: Ray Kim <raydhkim@amazon.com>

ZhenyeLee committed:
Add 'y' unit to four date functions (#549)

Previously DATEPART() and DATENAME() functions return incorrect result with 'y' unit. DATEDIFF() and DATEADD() returns correct values but unit 'y' is not the arguments of those two functions. This commit adds 'y' unit to DATEPART(), DATENAME(), DATEDIFF() and DATEADD()functions by adding 'y' as TSQL keyword.

Task: BABEL-2819
Signed-off-by: Zhenye Li zhenye@amazon.com

sumitj824 committed:
Add boundary conditions check for fixeddecimal/money/smallmoney datatypes when casting from bigint (#631)

Currently, Babelfish stores garbage values when the value exceeds the limit of fixeddecimal/money/smallmoney datatypes when casting from bigint. I have added boundary condition checks for fixeddecimal/money/smallmoney datatypes when casting from bigint so that will throw the error if the value exceeds the limit.

Signed-off-by: Sumit Jaiswal <sumiji@amazon.com>
Co-authored-by: Sumit Jaiswal <sumiji@amazon.com>

timchang514 committed:
[BABEL-3286] Support for FORMAT() (#548)

* [BABEL-3286] Support for FORMAT()

Signed-off-by: Tim Chang <changti@amazon.com>

* Fixed missing typo in format_exponential

* Fixed typo in cstring_to_numeric

* Typo with buf->data

* Fixed test diffs

* Fix possible heap corruption

* fix failing test, fix merge artifact

* whitespace

Signed-off-by: Tim Chang <changti@amazon.com>
Co-authored-by: Tim Chang <changti@amazon.com>

rossd-bitquill committed:
sys.default_constraints and sys.check_constraints updates to strip unexpected datatypes from definition column

This commit updates the definition column of sys.default_constraints and
sys.check_constraints to
1) update the datatype from varchar to nvarchar(4000) and
2) use a regex to strip the datatype from the definition column

Previously, the definition column used the result from pg_get_expr
directly; however, this function returns the datatype of non-numeric
expressions following  a double colon.  This is inconsistent with SQL
Server, which does not expect the inclusion of the datatype.

E.g. previously a varchar-type default of 'abc' would appear in
sys.default_constraints as 'abc'::'varchar'
After this change, the definitions will be corrected to 'abc'

Additionally, this commit adds tests for default and check constraints,
and upgrade tests for the same.

sys.checl_constraints has also had updates made to column types to match
definition in documentation.

Task: BABELFISH-524
Signed-off-by: rossd-bitquill <dutkiert@amazon.com>

timchang514 committed:
Add pg_catalog in front of format in upgrade script (#670)

Signed-off-by: Tim Chang <changti@amazon.com>

ZhenyeLee committed:
Fixed DELETE/UPDATE on correlation name with ANSI join raises error  (#595)

Previously, using DELETE/UPDATE on correlation name with ANSI join raises error "table name specified more than once”. This PR corrects tsql_update_delete_stmt_from_clause_alias which causes error. If user use INNER JOIN, the type of ListCell in from_clause List is JoinExpr instead of RangeVar Node type. However, the elements larg and rarg in ListCell is have info that we need to update relation. So if the input is in JoinExpr type, we will check if the aliasname is the same as left subtree or right subtree, then use corresponding subtree information to update relation.

Task:BABEL-3380
Signed-off-by: Zhenye Li zhenye@amazon.com

ZhenyeLee committed:
Correct 'nanosecond' unit action in DATEADD() function (#594)

Previously, DATEPART() supports the 'nanosecond' unit and returns the original datetime, which is not recognized in MSSQL. In the PR, nanosecond is accepted for DATETIMEOFFSET and DATETIME2 and TIME but not for DATETIME and SMALLDATETIME. In sys.dateadd_internal function, we check if the input datetype should support 'nanosecond' or not, and return corresponding return value or throw an error. Same as 'mircrosecond' unit. dateadd_internal_df function is also deleted to reduce code duplication, which is used to handle datetimeoffset input data type.

Task:BABEL-3474
Signed-off-by: Zhenye Li zhenye@amazon.com

rishabhtanwar29 committed:
Update upgrade test template (#650)

This commit does the following:

Adds upgrade paths for version 13.7 and 13.8. Currently tests for 13_6 are copied
to 13_7 and 13_8 upgrade directories.
Fixes branch template to use stable branches of extensions instead of tags.
Fixes branches in major/minor upgrade workflows for empty databases.
Task: BABEL-3583

Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>
Co-authored-by: Rishabh Tanwar <ritanwar@amazon.com>

AmithKamath committed:
Fix RAISERROR params limit issue

RAISERROR should not allow more than 20 arguments for the message
string. This commit adds a check to parse through the error message
string and count the number of format specifiers i.e the arguments.
An error is thrown if the number of arguments exceeds the limit i.e 20

The logic to count the number of format specifiers is tightly coupled
with the validation/preparation of error message string(prepare_format_string
in string.c)

This commit also adds JDBC tests for the changes mentioned above

Task: BABEL-2251

Signed-off-by: Amith Kamath <amithkam@amazon.com>

faviansamatha committed:
Added view definitions for sql_modules views  (#540)


Previously, sys.all_sql_modules, sys.sql_modules, and sys.system_modules returned null for view definitions due to Babelfish not storing these definitions in TSQL dialect. This commit fixes that issue. Furthermore, this commit also fixes the function and procedure definitions by displaying the definitions in TSQL-dialect rather than plpgsql.

Task: BABELFISH-529
Signed-off-by: Favian (Ian) Samatha ians@bitquilltech.com

faviansamatha committed:
Revert "Added view definitions for sql_modules views  (#540)" (#673)

This reverts commit 0408b1f9e8d73d05f2dab4ca5afa6f89679ee904.

wboettge committed:
Accessing a db without required permissions should not change the current db

Accessing a database with use without required permissions no longer
changes the current database. Previously, if a login attempted to access
a database they did not have permissions for, running SELECT DB_NAME()
after attepting to switch to the databaseb would return the database
they had been denied access from. Additionally, a shared lock would be
claimed, preventing other logins from deleting the database.

Now, the current_db session variables are unchanged and a lock is not
claimed following a failed attempt to access a database.

Task: BABEL-3585

Signed-off-by: Walt Boettge <wboettge@amazon.com>

jusjosj committed:
Support UDF as column default in ALTER TABLE (#581)

This change allows use of the syntax like

ALTER TABLE t2 ADD DEFAULT dbo.myf() FOR b

Previously, this type of statement would result in an error since we were not handling schema replacements in multidb.c for AT_ColumnDefault This PR adds logic to handle AT_ColumnDefault and calls the correct function to properly rewrite the dbo schema reference to its internal babelfish counter part

Task:BABEL-3268
Signed-off-by: Justin Jossick jusjosj@amazon.com

faviansamatha committed:
Added view definitions for sql_modules views (#674)

Previously, sys.all_sql_modules, sys.sql_modules, and sys.system_modules returned null for view definitions due to Babelfish not storing these definitions in TSQL dialect. This commit fixes that issue. Furthermore, this commit also fixes the function and procedure definitions by displaying the definitions in TSQL-dialect rather than plpgsql.

Task: BABELFISH-529
Signed-off-by: Favian (Ian) Samatha ians@bitquilltech.com

Commit:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3486

Commit:
Removed incorrect modification to all_sql_mods
Signed-off-by: Jake Owen <owjco@amazon.com>

Commit:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3486

Commit:
Updated test files with changes to comments

Signed-off-by: Jake Owen <owjco@amazon.com>

Commit:
Modified tests to remove unnecessary comments

Signed-off-by: Jake Owen <owjco@amazon.com>

Commit:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3486
